### PR TITLE
Implemented Timing user story, main engine, engine lifecycle.

### PR DESCRIPTION
**Pull request by gmalmquist**
Implemented the "Timing" user story. To do this I went ahead and created a "TrydentEngine" main class to run the engine, and along with a "ContinuousEvent" abstract class to represent events that the engine invokes each frame (which could be used by us for stuff like animations, or the users to actually program game behaviors based on time). And, of course, the Time class to keep track of time. 

Testing done with junits in LifecycleTest.java.

### CODE REVIEW

File: src/edu/gatech/cs2340/trydent/Time.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	 */
+	public static void setTimeRate(double rate) {
+		timeRate = rate;
+	}
+	
+	/**
+	 * Starts keeping track of time, called by the TrydentEngine.
+	 */
+	protected static void startTheDawnOfTime() {
+		realStartTime = System.currentTimeMillis();
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/Time.java]: While this is totally reasonable, do we instead perhaps want to use System.nanoTime()? As far as I am aware, these are identical except that nanoTime may have a higher resolution on some machines.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	private boolean started = false;
+	
+	public ContinuousEvent() {
+		TrydentEngine.addContinuousEvent(this);
+	}
+	
+	/**
+	 * Called by the engine on update.
+	 */
+	protected final void doUpdate() {
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Do we want protected or package access for this one? I think package access would make more sense, to rid any confusion between this method and onUpdate(). Unless you expect that we (or external users) would want to subclass ContinuousEvent and also override this method.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package edu.gatech.cs2340.trydent;
+
+/**
+ * Represents an event that is called by the engine every frame.
+ * @author Garrett Malmquist
+ *
+ */
+public abstract class ContinuousEvent {
+	
+	private boolean started = false;
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Spaces vs tabs? I guess we need to agree on some formatting rules, which is tangential to this pull.

File: src/edu/gatech/cs2340/trydent/Time.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	private static volatile double gameTimeSinceStartup;
+	private static volatile long realTimeSinceStartup;
+	private static volatile long realLastFrame;
+	private static volatile long realStartTime;
+	private static volatile double gameTimePassed;
+	private static volatile long realTimePassed;
+	private static volatile double timeRate;
+	
+	/**
+	 * Game time since the game started.
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/Time.java]: Add "in seconds" please.

File: src/edu/gatech/cs2340/trydent/Time.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	 * @return
+	 */
+	public static double getTime() {
+		return gameTimeSinceStartup / 1000.0;
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/Time.java]: Should we move 1000.0 into a constant? Do we care?

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+		continuousEventsToRemove.clear();
+		continuousEventsToAdd.clear();
+		
+		synchronized (mainLock) {
+			// We're done here.
+			mainThread = null;
+		}
+	}
+	
+	private static TrydentEngine getInstance() {
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: If you're implementing a Singleton, you should probably use an enum (exerpt from Effective Java: http://www.informit.com/articles/article.aspx?p=1216151&seqNum=3).

Alternatively, do we really want a singleton? Do you think some users might want to start 2 Trydent Engines at once? Or would being able to run multiple engines at once be convenient for testing? Or does that sound ridiculous?

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
+package edu.gatech.cs2340.trydent;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Main class that kicks off the TrydentEngine.
+ * 
+ * As a developer using the TrydentEngine, all you have to do is call
+ * TrydentEngine.start() to get started.
```
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java:10]: Calling MyCoolLibrary.start() seems like an antiquated way of starting up libraries. But maybe that's just my opinion.

Is there a better way of doing this? Like a static initializer block or lazy initialization? Or is this considered good practice?

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 1 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/Time.java]: I hear that nanoTime() is a slower function call than currentTimeMillis():

"nanoTime is significantly usually more accurate than currentTimeMillis but it's a relatively expensive call as well.currentTimeMillis() runs in a few (5-6) cpu clocks, nanoTime depends on the underlying architecture and can be 100+ cpu clocks. " - random stack overflow user
http://stackoverflow.com/questions/351565/system-currenttimemillis-vs-system-nanotime

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 2 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Good point, I'll change that.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 3 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Whoops. I need to  investigate my IDE settings. That's not supposed to be there =P

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 4 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/Time.java]: Yes.

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 1 above)
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/Time.java]: Well then I defer to random stack overflow user. keep milliseconds.

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 5 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/Time.java]: Mmmmm I guess it wouldn't hurt. I'm sure the compiler will optimize it away if I make it private static final.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 2 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: I can see users wanting to add Timer-like events what are contained within Trydent - leaving this open to subclassing is fine by me.

Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: I think this should probably have package scope because it seems to be an engine internal method. Also, this method can't be overridden because it is final.
Notably, whatever scope this has, it should probably be the same as doStop.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 3 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Spaces. I'm getting Gradle set up which should run our tests and checkstyle so we won't have to worry too hard about it in the future.

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 4 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/Time.java]: ++

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 6 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: Time() is already static; if we made it possible to make multiple TrydentEngines we'd have to correspondingly allow multiple corresponding Time's, which I think would just clutter the library for functionality no one really needs. Nothing stops them from running two instances of their whole program if they need to run things in parallel, if they're really motivated =P

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 3 above)
Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: I'm pretty sure we agreed to 4 spaces, 120 character lines, camel case variables, pascal case class names, spaced keywords, and Egyptian braces.

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 1 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/Time.java]: Internet seems to say that nanoTime is better wrt cross-platform performance, I'm all for it.
https://blogs.oracle.com/dholmes/entry/inside_the_hotspot_vm_clocks

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 5 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/Time.java]: Normally I'd say no but since the constant chances if we use nanoseconds it's probably a good idea.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 7 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java:10]: The trouble with static initializer blocks is that they'd only be activated at the point the JVM loaded in the TrydentEngine class; IE the first time TrydentEngine appears somewhere (directly or indirectly) in the end users' code. Lazy initialization in general has the same problem.

That seems really magic and weird and undesirable. It's running stuff in a separate thread, it needs to be started somehow. 

Also note that we might very well want to make start() take some parameters in the future, to set up the window or something.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
+package edu.gatech.cs2340.trydent;
+
+/**
+ * Represents an event that is called by the engine every frame.
+ * @author Garrett Malmquist
+ *
+ */
+public abstract class ContinuousEvent {
```
Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java:8]: Is this how the animations are going to be implemented internally?
Also, if the animation objects aren't going to be subclassing this, we may want to leverage the singleton or a dispatcher so that the users won't have to wrap everything into a ContinousEvent subclass.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 7 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java:10]: It's pretty antiquated but I could see an argument that it's the most easily understandable method of starting up the library. I think doing some static initializer block stuff might just be doing it differently for the sake of doing it differently and isn't really any better.

File: src/edu/gatech/cs2340/trydent/Time.java (hunk 5 above)
Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/Time.java]: We should care because we want this is for 2340, and the students should not come to think that magic numbers are okay.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 6 above)
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: Good point. This does seem like an extraneous use-case.

Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: If we want to make this a singleton, we should probably at least provide a mock class because testing is something that 2340 projects should be able to do easily.
Edit: On second thought, this will depend on how we want to manage game state.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 7 above)
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java:10]: What if a user forgot to call .start()? If we include a start() method, it seems like we should also litter our code with TrydentNotStartedException's. Which seems undesireable.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 2 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Aaron's misunderstanding of its purpose seems to be a poignant reason to make it package visibility =P

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 7 above)
Comment (djkeyes) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java:10]: IMO, feel free to leave it for now, but maybe we should discuss this in the future.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 6 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: I don't quite follow the need for a mock-class or any alternative instances. It's currently a singleton just because I thought it made the code a bit prettier internally, but that's not code that 2340 students need to look at. We could easily change it just to be a static class like Time; would that be preferable?

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 8 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java:8]: We might represent animations using that, I'm not sure yet. It's also a way for end users to actually hook into the engine in a thread-safe way. They could either just put their game logic in a single ContinuousEvent subclass in update, or subclass a lot of ContinuousEvents to represent different modules of their code (more elegantly, but also require more understanding of the library). This is similar to what Unity does with MonoBehaviors; I think it's the easiest way for users to interact with the engine in a threadsafe way.

Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java:8]: We should probably discuss this more regardless, because it's a key point of how our engine is structured.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 6 above)
Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: Actually, I think we can make the interfaces prettier if we use a singleton here, but maybe having some static methods would be useful for testing.
My rationale is that depending on what the event system is handling, a user may want to simulate the engine without introducing state into the singleton to test edge cases.

Edit: Also, unit tests ideally would not involve real memory or graphics resources.

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 8 above)
Comment (winterfroststrom) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java:8]: If that's the case, is there a reason to call it a ContinuousEvent? Why not just TrydentEvent? Or even TrydentBehavior?

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 2 above)
Comment (AaronFriesen) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java]: Missed the final. Blarg. Yeah package sounds good

File: src/edu/gatech/cs2340/trydent/ContinuousEvent.java (hunk 8 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/ContinuousEvent.java:8]: I wasn't sure what to call it, so I asked Aaron and that's what he said. Renaming is super easy to do later though with automatic refactoring tools.

The idea of "ContinuousEvent" is it's an event that occurs repeatedly - continuously - every frame.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 7 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java:10]: So a benefit of TrydentEngine being a singleton is that anytime a user interacts with it, getInstance() is called internally. Which means we could easily automatically call start() there if we wanted to invoke it lazily in the future. So perhaps we should leave it as-is for now, and also leave TrydentEngine as a singleton so that we can easily add that later.

File: src/edu/gatech/cs2340/trydent/TrydentEngine.java (hunk 6 above)
Comment (gmalmquist) [inline src/edu/gatech/cs2340/trydent/TrydentEngine.java]: There already are passing tests for the engine lifecycle that work fine without any memory or graphics stuff.
